src/librbd/io/AioCompletion.cc

```diff
@@ -273,11 +275,11 @@ void AioCompletion::complete_event_socket() {
 }
 
 void AioCompletion::notify_callbacks_complete() {
-  state = AIO_STATE_COMPLETE;
+  state.store(AIO_STATE_COMPLETE, std::memory_order_release);
 
-  {
+  if (cond != nullptr && waiting_for_complete.load(std::memory_order_acquire)) {
```

Comment:
Is accessing `cond` without holding the mutex safe?  Can the assignment to `cond` in `wait_for_complete()` be reordered with the initial test on `state`/become visible to the thread executing this condition late, so that the thread executing `wait_for_complete()` ends up waiting on `cond` but we never notify it because `cond` happened to test NULL?

Reply:
> Can the assignment to `cond` in `wait_for_complete()` be reordered with the initial test on `state`/become visible to the thread executing this condition late

If not, wouldn't it be reported as a data race anyway?

Reply:
Good point about the race hazard. Would you recommend moving the nullptr check to after the mutex, or removing it entirely?

I don't think it's possible for cond to be null if waiting_for_complete is true, but I thought it would be good practice to add a nullptr check anyway.

Reply:
> Would you recommend moving the nullptr check to after the mutex, or removing it entirely?

With the code as is, the nullptr check can't removed entirely because of another race.  Consider:
- `wait_for_complete()` sets `waiting_for_complete` to true but doesn't get a change to grab the mutex
- `notify_callbacks_complete()` loads `waiting_for_complete`, sees that it's true, grabs the mutex and attempts to call `notify_all()` on `cond` that hasn't even been allocated, let alone "published"

I'm hoping we can get rid of all of this with the switch to C++20 atomic wait/notify.

Reply:
> I don't think it's possible for cond to be null if waiting_for_complete is true

It seems possible to me, as a result of the race that I outlined above.

Reply:
Yes, sorry - I meant not possible under normal circumstances, but obviously it is possible in the race condition.

Good point that we won't need the mutex with the C++20 atomic wait/notify so I will make that change first.